### PR TITLE
♻️ Modernize polyfills

### PR DESCRIPTION
**Pull request by rcebulko**
Part of #34349

### CODE REVIEW

File: src/polyfills/custom-elements.js [hunk 1]
```diff
@@ -491,9 +479,9 @@ class Registry {
     // I would love to not have to hold onto all of the roots, since it's a
     // memory leak. Unfortunately, there's no way to iterate a list and hold
     // onto its contents weakly.
-    this.roots_.forEach((tree) => {
+    for (const tree of this.roots_) {
```
Comment (rcebulko) [inline src/polyfills/custom-elements.js:482]: Does iterator here resolve the issue in the above comment at all? Not clear on what "hold onto its contents weakly" means

Comment (jridgewell) [reply src/polyfills/custom-elements.js:482]: No, we need a `WeakSet` here, but you can't iterate a `WeakSet`.

File: src/polyfills/fetch.js [hunk 2]
```diff
@@ -68,9 +68,9 @@ export function fetchPolyfill(input, init = {}) {
     }
 
     if (init.headers) {
-      Object.keys(init.headers).forEach(function (header) {
+      for (const header of Object.keys(init.headers)) {
```
Comment (jridgewell) [inline src/polyfills/fetch.js:71]: Fetch polyfill is explicitly for IE11, so let's undo this.

File: src/polyfills/fetch.js [hunk 3]
```diff
@@ -280,12 +280,12 @@ export class Response extends FetchResponse {
     data.status = init.status === undefined ? 200 : parseInt(init.status, 10);
 
     if (isArray(init.headers)) {
-      /** @type {!Array} */ (init.headers).forEach((entry) => {
+      for (const entry of /** @type {!Array} */ (init.headers)) {
```
Comment (jridgewell) [inline src/polyfills/fetch.js:283]: Ditto.

File: src/polyfills/object-assign.js [hunk 4]
```diff
@@ -30,8 +30,7 @@ export function assign(target, var_args) {
   }
 
   const output = Object(target);
-  for (let i = 1; i < arguments.length; i++) {
-    const source = arguments[i];
+  for (const source of arguments) {
```
Comment (jridgewell) [inline src/polyfills/object-assign.js:33]: Ditto explicitly for IE11.

File: src/polyfills/stubs/intersection-observer-stub.js [hunk 5]
```diff
@@ -268,7 +268,9 @@ export class IntersectionObserverStub {
   upgrade_(Ctor) {
     const inst = new Ctor(this.callback_, this.options_);
     this.inst_ = inst;
-    this.elements_.forEach((e) => inst.observe(e));
+    for (const e of /** @type {!Array<!Element>} */ (this.elements_)) {
```
Comment (jridgewell) [inline src/polyfills/stubs/intersection-observer-stub.js:271]: Nit: This is kinda ugly. I would just remove the nullability, and instead truncate arrays when clearing memory.

File: src/polyfills/stubs/resize-observer-stub.js [hunk 6]
```diff
@@ -144,7 +146,9 @@ export class ResizeObserverStub {
   upgrade_(Ctor) {
     const inst = new Ctor(this.callback_);
     this.inst_ = inst;
-    this.elements_.forEach((e) => inst.observe(e));
+    for (const e of /** @type {!Array<!Element>} */ (this.elements_)) {
```
Comment (jridgewell) [inline src/polyfills/stubs/resize-observer-stub.js:149]: Nit: Ditto nullibility.

**Review by jridgewell — APPROVED** [review]

File: src/polyfills/fetch.js (hunk 2 above)
Comment (rcebulko) [reply src/polyfills/fetch.js]: Reverted this file.

File: src/polyfills/fetch.js (hunk 3 above)
Comment (rcebulko) [reply src/polyfills/fetch.js]: Done

File: src/polyfills/object-assign.js (hunk 4 above)
Comment (rcebulko) [reply src/polyfills/object-assign.js]: ditto reverted

File: src/polyfills/stubs/intersection-observer-stub.js (hunk 5 above)
Comment (rcebulko) [reply src/polyfills/stubs/intersection-observer-stub.js]: Done via `this.elements_.length = 0` following convention above.

File: src/polyfills/stubs/resize-observer-stub.js (hunk 6 above)
Comment (rcebulko) [reply src/polyfills/stubs/resize-observer-stub.js]: Done.

**Review by jridgewell — APPROVED** [review]

**Pull request merged**